src/windows_uwp.rs

```diff
@@ -0,0 +1,42 @@
+// Copyright 2018 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation for Windows UWP targets. After deprecation of Windows XP
+//! and Vista, this can superseed the `RtlGenRandom`-based implementation.
+use crate::Error;
+use core::{ffi::c_void, num::NonZeroU32, ptr};
+
+const BCRYPT_USE_SYSTEM_PREFERRED_RNG: u32 = 0x00000002;
+
+extern "system" {
+    fn BCryptGenRandom(
+        hAlgorithm: *mut c_void,
+        pBuffer: *mut u8,
+        cbBuffer: u32,
+        dwFlags: u32,
+    ) -> i32;
+}
+
+pub fn getrandom_inner(dest: &mut [u8]) -> Result<(), Error> {
+    // Prevent overflow of u32
+    for chunk in dest.chunks_mut(u32::max_value() as usize) {
+        let ret = unsafe {
+            BCryptGenRandom(
+                ptr::null_mut(),
+                chunk.as_mut_ptr(),
+                chunk.len() as u32,
+                BCRYPT_USE_SYSTEM_PREFERRED_RNG,
+            )
+        };
+        if let Some(code) = NonZeroU32::new(ret as u32) {
```

Comment:
My main concern with this is that [`NSSTATUS` values use the entire `u32` space](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/87fba13e-bf06-450e-83b1-9241dc81e781). This means that success/informational statuses will look like OS errors, warning statuses will look like internal errors, and error statuses will look like custom errors. This is not ideal.

We should either:
  - Be more careful with how we map `NTSTATUS` to `Error`
  - Just use a single internal return value, like how we use `RTL_GEN_RANDOM_FAILED` for the normal Windows implementation.

Reply:
Hm, this is really unfortunate... I am not sure if erasing information completely is better than using a lossy mapping. So one option is to simply zeroize the highest bit.

According to `BCryptGenRandom` [docs](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom), it should return `STATUS_SUCCESS` (i.e. 0) if call was successful. To be safe I think it will be better to be conservative and interpret all other codes as errors, even if they were informational or warning.